README.md
# AFMActionSheet

[![CI Status](http://img.shields.io/travis/ask-fm/AFMActionSheet.svg?style=flat)](https://travis-ci.org/ask-fm/AFMActionSheet)
[![Version](https://img.shields.io/cocoapods/v/AFMActionSheet.svg?style=flat)](http://cocoapods.org/pods/AFMActionSheet)
[![License](https://img.shields.io/cocoapods/l/AFMActionSheet.svg?style=flat)](http://cocoapods.org/pods/AFMActionSheet)
[![Platform](https://img.shields.io/cocoapods/p/AFMActionSheet.svg?style=flat)](http://cocoapods.org/pods/AFMActionSheet)

AFMActionSheet provides a AFMActionSheetController that can be used in places where one would use a UIAlertController, but a customized apperance or custom presentation/dismissal animation is needed. Seeing as how AFMActionSheetController was inspired by UIAlertController, it too supports ActionSheet and Alert styles to make your life even easier.

![Action Sheet Example](https://raw.githubusercontent.com/ask-fm/AFMActionSheet/master/res/action_sheet.gif)
![Alert Example](https://raw.githubusercontent.com/ask-fm/AFMActionSheet/master/res/alert.gif)

## Usage
To create an action sheet with default style and default transition animations:
```swift
let actionSheet = AFMActionSheetController()
let action = AFMAction(title: "Action", enabled: true, handler: { (action: AFMAction) -> Void in
    // Do something in handler
}
actionSheet.add(action)
self.present(actionSheet, animated: true, completion: {
    // Do something after completion
})
```
That's it.

## Detailed look
### Style
AFMActionSheetController supports two styles: the default action sheet style and alert style. These are set via initializers with `ControllertStyle` enum (`ControllertStyle.actionSheet` and `ControllertStyle.alert` accordingly)
```swift
let actionSheet = AFMActionSheetController(style: .actionSheet)
let alert = AFMActionSheetController(style: .alert)
```

### Transitioning animations
To change presentation and dismissal animations, implement a `UIViewControllerTransitioningDelegate` and pass it to AFMActionSheetController
```swift
let actionSheet = AFMActionSheetController(transitioningDelegate: myCustomTransitioningDelegate)
```
or
```swift
actionSheet.setup(myCustomTransitioningDelegate)
```

### Action controls
Action sheet's controls are created by adding `AFMAction` objects to the controller
```swift
actionSheet.add(action)
```
It is also possible to add actions as "Cancel actions". When using `ControllertStyle.ActionSheet` style, controls for these actions will be displayed in the bottom "Cancel section".
```swift
actionSheet.add(cancelling: action)
```
To use custom views as action controls just pass the view with action to `add(_:with:)` or `add(cancelling:with:)` method
```swift
actionSheet.add(action, with: myCustomControl)
actionSheet.add(cancelling: action, with: myCustomControl)
```
Height of the action control is whatever height passed custom view specifies, but it is possible to specify minimal control height with `minControlHeight` property.

### Title view
Title view is a view that is located on top of the action sheet and works similar to action controls.
To set the title view with custom view
```swift
actionSheet.add(title: myCustomTitleView)
```
or to set a default `UILabel` with a text
```swift
actionSheet.add(titleLabelWith: "Title")
```
Like with action control height, it is possible to specify minimal control height with `minTitleHeight` property.

### Other customizations
There is a number of properties to help further modify the look and behavior of action sheet controller:

- `spacing: Int` specifies content's spacing between action controls (default is `4`)
- `horizontalMargin: Int` specifies spacing from content to controller's top and bottom (default is `16`)
- `verticalMargin: Int` specifies spacing from content to controller's  left and right (default is `16`)
- `cornerRadius: Int` specifies corner radius of content (default is `10`)
- `backgroundColor: UIColor` specifies controller's background color (default is `blackColor().colorWithAlphaComponent(0.5)`)
- `spacingColor: UIColor` specifies content's spacing color (default is `.clearColor()`)
- `outsideGestureShouldDismiss: Bool` specifies whether the click on background outside of controls and title dismisses action sheet (default is `true`)


## Example project
To run the example project, clone the repo, and run `pod install` from the Example directory first.

## Requirements
iOS 8 and up.

## Installation

AFMActionSheet is available through [CocoaPods](http://cocoapods.org). To install
it, simply add the following line to your Podfile:

```ruby
pod "AFMActionSheet"
```

## Contact person

Ilya Alesker, ilya.alesker@ask.fm

## License

AFMActionSheet is available under the MIT license. See the LICENSE file for more info.


CHANGELOG.md
<!---
[x.x.x] - yyyy-MM-dd
=============================================================

### New:

* None

### Bugfixes:

* None

### Technical

* None
-->


[2.0.1] - 2018-01-31
=============================================================

### New:

* Swift 4

[2.0.0] - 2016-10-20
=============================================================

### New:

* Swift 3


[1.1.0] - 2016-10-12
=============================================================

This release contains several enhancements and improvements

### New:

* Added an option to insert action at position
* Added dismiss completion block

### Bugfixes:

* Added vertical constraints for Alert style

### Technical

* Updated project for cocoapods 1.0.1


[1.0.0] - 2016-03-30
=============================================================

Initial release

<!---
[x.x.x]: https://github.com/ask-fm/AFMActionSheet/releases/tag/x.x.x
-->
[2.0.0]: https://github.com/ask-fm/AFMActionSheet/releases/tag/2.0.0
[1.1.0]: https://github.com/ask-fm/AFMActionSheet/releases/tag/1.1.0
[1.0.0]: https://github.com/ask-fm/AFMActionSheet/releases/tag/1.0.0


Example/Tests/Tests.swift
// https://github.com/Quick/Quick

import Quick
import Nimble
import AFMActionSheet

class TableOfContentsSpec: QuickSpec {
    override func spec() {
        describe("these will fail") {

            it("can do maths") {
                expect(1) == 2
            }

            it("can read") {
                expect("number") == "string"
            }

            it("will eventually fail") {
                expect("time").toEventually( equal("done") )
            }
            
            context("these will pass") {

                it("can do maths") {
                    expect(23) == 23
                }

                it("can read") {
                    expect("🐮") == "🐮"
                }

                it("will eventually pass") {
                    var time = "passing"

                    dispatch_async(dispatch_get_main_queue()) {
                        time = "done"
                    }

                    waitUntil { done in
                        NSThread.sleepForTimeInterval(0.5)
                        expect(time) == "done"

                        done()
                    }
                }
            }
        }
    }
}


Example/Tests/ActionSheetControllerSpec.swift
// https://github.com/Quick/Quick

import Quick
import Nimble
import Nimble_Snapshots
@testable import AFMActionSheet

class ActionSheetControllerSpec: QuickSpec {
    override func spec() {
        describe("created action sheet controller") {
            it("should have correct transition delegate, correct presentation style and controller style") {
                let controller = AFMActionSheetController(transitioningDelegate: AFMActionSheetTransitioningDelegate())
                expect(controller.transitioningDelegate).toNot(beNil())
                expect(controller.modalPresentationStyle).to(equal(UIModalPresentationStyle.custom))
                expect(controller.controllerStyle).to(equal(AFMActionSheetController.ControllerStyle.actionSheet))
            }

            it("should have gesture recognizers set up") {
                let controller = AFMActionSheetController()
                expect(controller.view.gestureRecognizers!.count).to(equal(2))
                expect(controller.view.gestureRecognizers!.first as? UISwipeGestureRecognizer).toNot(beNil())
                expect(controller.view.gestureRecognizers!.last as? UITapGestureRecognizer).toNot(beNil())
            }
        }

        describe("basic controller setup") {
            var controller: AFMActionSheetController!
            var action : AFMAction!

            beforeEach {
                controller = MockActionSheetController()
                action = AFMAction(title: "action", enabled: true, handler: nil)
            }

            it("should have correct number of actions and subviews when adding 1 action") {
                controller.add(action)
                expect(controller.actions.count).to(equal(1))
                expect(controller.actionControls.count).to(equal(1))
                expect(controller.actionGroupView.subviews.count).to(equal(1))
            }

            it("should have correct number of actions and subviews when adding 2 actions") {
                controller.add(action)
                controller.add(action)
                expect(controller.actions.count).to(equal(2))
                expect(controller.actionControls.count).to(equal(2))
                expect(controller.actionGroupView.subviews.count).to(equal(2))
            }

            it("should have correct number of actions and subviews when inserting an action") {
                controller.add(action)
                controller.add(action)
                let insertedControl = UIButton.control(with: action)
                controller.insert(action, with: insertedControl, at: 0)
                expect(controller.actions.count).to(equal(3))
                expect(controller.actionControls.count).to(equal(3))
                expect(controller.actionControls.first).to(equal(insertedControl))
                expect(controller.actionGroupView.subviews.count).to(equal(3))
            }

            it("should have correct number of actions and subviews when inserting an action at wrong position") {
                controller.add(action)
                controller.add(action)
                let insertedControl = UIButton.control(with: action)
                controller.insert(action, with: insertedControl, at: 42)
                expect(controller.actions.count).to(equal(2))
                expect(controller.actionControls.count).to(equal(2))
                expect(controller.actionControls.first).toNot(equal(insertedControl))
                expect(controller.actionGroupView.subviews.count).to(equal(2))
            }

            it("should have correct number of subviews when adding title view") {
                let titleView = UIView()
                controller.add(title: titleView)
                expect(controller.actionGroupView.subviews.count).to(equal(1))
            }

            it("should have label as title view view when adding title") {
                controller.add(titleLabelWith: "title")
                expect(controller.titleView as? UILabel).toNot(beNil())
                expect((controller.titleView as! UILabel).text).to(equal("title"))
                expect(controller.actionGroupView.subviews.count).to(equal(1))
            }

            it("should have control enabled if added action was enabled") {
                action.enabled = true
                controller.add(action)
                expect(controller.actionControls.first?.isEnabled).to(beTrue())
            }

            it("should have control disabled if added action was disabled") {
                action.enabled = false
                controller.add(action)
                expect(controller.actionControls.first?.isEnabled).to(beFalse())
            }

            it("should call correct action's handler when tapping on correspoding control") {
                var calledAction1 = false
                let action1 = AFMAction(title: "action1", enabled: true) { _ in
                    calledAction1 = true
                }
                var calledAction2 = false
                let action2 = AFMAction(title: "action2", enabled: true) { _ in
                    calledAction2 = true
                }
                controller.add(action1)
                controller.add(action2)
                controller.handleTaps(controller.actionControls.first!)
                expect(calledAction1).to(beTrue())
                expect(calledAction2).to(beFalse())
            }

            it("should not call action's handler if action was disabled") {
                var calledAction1 = false
                let action1 = AFMAction(title: "action1", enabled: false) { _ in
                    calledAction1 = true
                }
                controller.add(action1)
                controller.handleTaps(controller.actionControls.last!)
                expect(calledAction1).to(beFalse())
            }
        }

        describe("action sheet controller setup") {
            var controller: AFMActionSheetController!
            var action : AFMAction!

            beforeEach {
                controller = MockActionSheetController()
                action = AFMAction(title: "action", enabled: true, handler: nil)
            }

            it("should have set up action and cancel group views correctly") {
                expect(controller.actionGroupView.superview == controller.view).to(beTrue())
                expect(controller.cancelGroupView.superview == controller.view).to(beTrue())
            }

            it("should have correct number of actions and subviews when adding 1 cancel action") {
                controller.add(cancelling: action)
                expect(controller.actions.count).to(equal(1))
                expect(controller.cancelControls.count).to(equal(1))
                expect(controller.cancelGroupView.subviews.count).to(equal(1))
            }

            it("should have correct number of actions and subviews when adding different actions") {
                controller.add(action)
                controller.add(cancelling: action)
                controller.add(cancelling: action)
                controller.add(action)
                expect(controller.actions.count).to(equal(4))
                expect(controller.actionControls.count).to(equal(2))
                expect(controller.cancelControls.count).to(equal(2))
                expect(controller.actionGroupView.subviews.count).to(equal(2))
                expect(controller.cancelGroupView.subviews.count).to(equal(2))
            }
        }

        describe("alert controller setup") {
            var controller: AFMActionSheetController!
            var action : AFMAction!

            beforeEach {
                controller = MockActionSheetController(style: .alert)
                action = AFMAction(title: "action", enabled: true, handler: nil)
            }

            it("should have set up action and cancel group views correctly") {
                expect(controller.actionGroupView.superview == controller.view).to(beTrue())
                expect(controller.cancelGroupView.superview == nil).to(beTrue())
            }

            it("should have correct number of actions and subviews when adding 1 cancel action") {
                controller.add(cancelling: action)
                expect(controller.actions.count).to(equal(1))
                expect(controller.cancelControls.count).to(equal(1))
                expect(controller.cancelGroupView.subviews.count).to(equal(0))
            }

            it("should have correct number of actions and subviews when adding different actions") {
                controller.add(action)
                controller.add(cancelling: action)
                controller.add(cancelling: action)
                controller.add(action)
                expect(controller.actions.count).to(equal(4))
                expect(controller.actionControls.count).to(equal(2))
                expect(controller.cancelControls.count).to(equal(2))
                expect(controller.actionGroupView.subviews.count).to(equal(4))
                expect(controller.cancelGroupView.subviews.count).to(equal(0))
            }
        }

        describe("AFMActionSheetController", {
            it("has valid snapshot for action sheet setup") {
                let controller = AFMActionSheetController(style: .actionSheet, transitioningDelegate: AFMActionSheetTransitioningDelegate())
                self.setup(controller: controller)
                expect(controller.view).to(haveValidSnapshot(named: "AFMActionSheetController-ActionSheet-Snapshot"))
            }

            it("has valid snapshot for alert setup") {
                let controller = AFMActionSheetController(style: .alert, transitioningDelegate: AFMActionSheetTransitioningDelegate())
                self.setup(controller: controller)
                expect(controller.view).to(haveValidSnapshot(named: "AFMActionSheetController-Alert-Snapshot"))
            }
        });
    }

    func setup(controller: AFMActionSheetController) {
        let action1 = AFMAction(title: "Action 1", enabled: true, handler: nil)
        let action2 = AFMAction(title: "Action 2", enabled: false, handler: nil)
        let action3 = AFMAction(title: "Action 3", handler: nil)

        controller.add(action1)
        controller.add(action2)
        controller.add(cancelling: action3)
        controller.add(titleLabelWith: "Title")
    }

    class MockActionSheetController : AFMActionSheetController {
        override func dismiss(animated flag: Bool, completion: (() -> Void)?) {
            completion?()
        }
    }
}


Pod/Classes/AFMActionSheetController.swift
//
//  ActionSheetViewController.swift
//  askfm
//
//  Created by Ilya Alesker on 26/08/15.
//  Copyright (c) 2015 Ask.fm Europe, Ltd. All rights reserved.
//

import UIKit

@IBDesignable
open class AFMActionSheetController: UIViewController {

    public enum ControllerStyle : Int {
        case actionSheet
        case alert
    }

    @IBInspectable open var outsideGestureShouldDismiss: Bool = true

    @IBInspectable open var minControlHeight: Int     = 50 {
        didSet { self.updateUI() }
    }
    @IBInspectable open var minTitleHeight: Int       = 50 {
        didSet { self.updateUI() }
    }
    @IBInspectable open var spacing: Int              = 4  {
        didSet { self.updateUI() }
    }
    @IBInspectable open var horizontalMargin: Int     = 16 {
        didSet { self.updateUI() }
    }
    @IBInspectable open var verticalMargin: Int       = 16 {
        didSet { self.updateUI() }
    }
    @IBInspectable open var cornerRadius: Int         = 10 {
        didSet { self.updateUI() }
    }

    @IBInspectable open var backgroundColor: UIColor = UIColor.black.withAlphaComponent(0.5) {
        didSet { self.updateUI() }
    }

    @IBInspectable open var spacingColor: UIColor = UIColor.clear {
        didSet { self.updateUI() }
    }

    private var topMargin: Int {
        let statusBarHeight = Int(UIApplication.shared.statusBarFrame.height)
        return statusBarHeight > self.verticalMargin ? statusBarHeight : self.verticalMargin
    }

    open var dismissCompletionBlock: (() -> ())?

    let controllerStyle: ControllerStyle

    public private(set) var actions: [AFMAction] = []
    public private(set) var actionControls: [UIControl] = []
    public private(set) var cancelControls: [UIControl] = []
    public private(set) var titleView: UIView?

    private var actionControlConstraints: [NSLayoutConstraint] = []
    private var cancelControlConstraints: [NSLayoutConstraint] = []

    private var actionSheetTransitioningDelegate: UIViewControllerTransitioningDelegate?

    var actionGroupView: UIView = UIView()
    var cancelGroupView: UIView = UIView()


    // MARK: Initializers

    public init(style: ControllerStyle, transitioningDelegate: UIViewControllerTransitioningDelegate) {
        self.controllerStyle = style
        super.init(nibName: nil, bundle: nil)
        self.setupViews()
        self.setup(transitioningDelegate)
    }

    public convenience init(style: ControllerStyle) {
        self.init(style: style, transitioningDelegate: AFMActionSheetTransitioningDelegate())
    }

    public convenience init(transitioningDelegate: UIViewControllerTransitioningDelegate) {
        self.init(style: .actionSheet, transitioningDelegate: transitioningDelegate)
    }

    override init(nibName nibNameOrNil: String?, bundle nibBundleOrNil: Bundle?) {
        self.controllerStyle = .actionSheet
        super.init(nibName: nibNameOrNil, bundle: nibBundleOrNil)
        self.setupViews()
        self.setup(AFMActionSheetTransitioningDelegate())
    }

    required public init?(coder aDecoder: NSCoder) {
        self.controllerStyle = .actionSheet
        super.init(coder: aDecoder)
        self.setupViews()
        self.setup(AFMActionSheetTransitioningDelegate())

    }

    private func setupViews() {
        self.setupGroupViews()
        self.setupGestureRecognizers()

        self.view.backgroundColor = self.backgroundColor
    }

    private func setupGestureRecognizers() {
        if let gestureRecognizers = self.view.gestureRecognizers {
            for gestureRecognizer in gestureRecognizers {
                self.view.removeGestureRecognizer(gestureRecognizer)
            }
        }

        let swipeGestureRecognizer = UISwipeGestureRecognizer(target: self, action: #selector(AFMActionSheetController.recognizeGestures(_:)))
        swipeGestureRecognizer.direction = UISwipeGestureRecognizerDirection.down
        let tapGestureRecognizer = UITapGestureRecognizer(target: self, action: #selector(AFMActionSheetController.recognizeGestures(_:)))
        tapGestureRecognizer.cancelsTouchesInView = false

        self.view.addGestureRecognizer(swipeGestureRecognizer)
        self.view.addGestureRecognizer(tapGestureRecognizer)
    }

    open func setup(_ transitioningDelegate: UIViewControllerTransitioningDelegate) {
        self.modalPresentationStyle = .custom
        self.actionSheetTransitioningDelegate = transitioningDelegate
        self.transitioningDelegate = self.actionSheetTransitioningDelegate
    }


    // MARK: Adding actions

    private func setup(_ control: UIControl, with action: AFMAction) {
        control.isEnabled = action.enabled
        control.addTarget(self, action:#selector(AFMActionSheetController.handleTaps(_:)), for: .touchUpInside)
        control.tag = self.actions.count
    }

    @objc
    open func add(_ action: AFMAction, with control: UIControl, andActionIsCancelling isCancellingAction: Bool) {
        self.setup(control, with: action)
        self.actions.append(action)

        if isCancellingAction {
            self.cancelControls.append(control)
        } else {
            self.actionControls.append(control)
        }

        self.addToGroupView(control, andActionIsCancelling: isCancellingAction)
    }

    @objc
    open func insert(_ action: AFMAction, with control: UIControl, at position: Int, andActionIsCancelling isCancellingAction: Bool) {
        if isCancellingAction {
            guard position <= self.cancelControls.count else { return }
        } else {
            guard position <= self.actionControls.count else { return }
        }
        self.setup(control, with: action)
        self.actions.append(action)

        if isCancellingAction {
            self.cancelControls.insert(control, at: position)
        } else {
            self.actionControls.insert(control, at: position)
        }

        self.addToGroupView(control, andActionIsCancelling: isCancellingAction)
    }

    @objc
    open func add(title: UIView) {
        self.titleView = title

        self.titleView!.translatesAutoresizingMaskIntoConstraints = false
        self.actionGroupView.addSubview(self.titleView!)
        self.updateContraints()
    }


    private func addToGroupView(_ control: UIControl, andActionIsCancelling isCancellingAction: Bool) {
        if self.controllerStyle == .actionSheet {
            self.addToGroupViewForActionSheet(control, andActionIsCancelling: isCancellingAction)
        } else if self.controllerStyle == .alert {
            self.addToGroupViewForAlert(control, andActionIsCancelling: isCancellingAction)
        }
    }

    private func addToGroupViewForActionSheet(_ control: UIControl, andActionIsCancelling isCancellingAction: Bool) {
        control.translatesAutoresizingMaskIntoConstraints = false
        if isCancellingAction {
            self.cancelGroupView.addSubview(control)
        } else {
            self.actionGroupView.addSubview(control)
        }
        self.updateContraints()
    }

    private func addToGroupViewForAlert(_ control: UIControl, andActionIsCancelling isCancellingAction: Bool) {
        control.translatesAutoresizingMaskIntoConstraints = false
        self.actionGroupView.addSubview(control)
        self.updateContraints()
    }

    private func actionControlsWithTitle() -> [UIView] {
        var views: [UIView] = self.actionControls
        if let titleView = self.titleView {
            views.insert(titleView, at: 0)
        }
        return views
    }


    // MARK: Control positioning and updating

    func updateContraints() {
        if self.controllerStyle == .actionSheet {
            self.updateContraintsForActionSheet()
        } else if self.controllerStyle == .alert {
            self.updateContraintsForAlert()
        }
    }

    func updateContraintsForActionSheet() {
        self.cancelGroupView.removeConstraints(self.cancelControlConstraints)
        self.cancelControlConstraints = self.constraints(for: self.cancelControls)
        self.cancelGroupView.addConstraints(self.cancelControlConstraints)

        self.actionGroupView.removeConstraints(self.actionControlConstraints)
        self.actionControlConstraints = self.constraints(for: self.actionControlsWithTitle())
        self.actionGroupView.addConstraints(self.actionControlConstraints)
    }

    func updateContraintsForAlert() {
        var views: [UIView] = self.actionControlsWithTitle()
        let cancelViews: [UIView] = self.cancelControls
        views.append(contentsOf: cancelViews)
        self.actionGroupView.removeConstraints(self.actionControlConstraints)
        self.actionControlConstraints = self.constraints(for: views)
        self.actionGroupView.addConstraints(self.actionControlConstraints)
    }

    private func setupGroupViews() {
        if self.controllerStyle == .actionSheet {
            self.setupGroupViewsForActionSheet()
        } else if self.controllerStyle == .alert {
            self.setupGroupViewsForAlert()
        }
        self.actionGroupView.backgroundColor = self.spacingColor
        self.cancelGroupView.backgroundColor = self.spacingColor
    }

    private func setupGroupViewsForActionSheet() {
        let setupGroupView: (UIView) -> Void = { groupView in
            groupView.removeFromSuperview()
            self.view.addSubview(groupView)
            groupView.clipsToBounds = true
            groupView.layer.cornerRadius = CGFloat(self.cornerRadius)
            groupView.translatesAutoresizingMaskIntoConstraints = false
            self.view.addConstraints(NSLayoutConstraint.constraints(withVisualFormat: "H:|-margin-[groupView]-margin-|",
                options: NSLayoutFormatOptions(),
                metrics: ["margin": self.horizontalMargin],
                views: ["groupView": groupView])
            )
        }

        setupGroupView(self.actionGroupView)
        setupGroupView(self.cancelGroupView)

        self.view.addConstraints(NSLayoutConstraint.constraints(withVisualFormat: "V:|-(>=topMargin)-[actionGroupView]-margin-[cancelGroupView]-margin-|",
            options: NSLayoutFormatOptions(),
            metrics: ["topMargin": self.topMargin, "margin": self.verticalMargin],
            views: ["actionGroupView": self.actionGroupView, "cancelGroupView": self.cancelGroupView])
        )
    }

    private func setupGroupViewsForAlert() {
        self.actionGroupView.removeFromSuperview()
        self.view.addSubview(self.actionGroupView)

        self.actionGroupView.clipsToBounds = true
        self.actionGroupView.layer.cornerRadius = CGFloat(self.cornerRadius)
        self.actionGroupView.translatesAutoresizingMaskIntoConstraints = false
        self.view.addConstraints(NSLayoutConstraint.constraints(withVisualFormat: "H:|-margin-[groupView]-margin-|",
            options: NSLayoutFormatOptions(),
            metrics: ["margin": self.horizontalMargin],
            views: ["groupView": self.actionGroupView])
        )

        self.view.addConstraint(NSLayoutConstraint(item: self.actionGroupView,
            attribute: .centerY,
            relatedBy: .equal,
            toItem: self.view,
            attribute: .centerY,
            multiplier: 1,
            constant: 0))

        self.view.addConstraints(NSLayoutConstraint.constraints(withVisualFormat: "V:|-(>=margin)-[actionGroupView]-(>=margin)-|",
            options: NSLayoutFormatOptions(),
            metrics: ["margin": self.topMargin],
            views: ["actionGroupView": self.actionGroupView]))
    }

    private func constraints(for views: [UIView]) -> [NSLayoutConstraint] {
        var constraints: [NSLayoutConstraint] = []

        var sibling: UIView?
        // we want to position controls from top to bottom
        for view in views.reversed() {
            let isLast = view == views.first
            constraints.append(contentsOf: self.horizontalConstraints(for: view))
            constraints.append(contentsOf: self.verticalConstraints(for: view, withSibling: sibling, andIsLast: isLast))

            sibling = view
        }
        
        return constraints
    }

    private func horizontalConstraints(for view: UIView) -> [NSLayoutConstraint] {
        return NSLayoutConstraint.constraints(withVisualFormat: "H:|-0-[view]-0-|",
            options: NSLayoutFormatOptions(),
            metrics: nil,
            views: ["view": view])
    }

    private func verticalConstraints(for view: UIView, withSibling sibling: UIView?, andIsLast isLast: Bool) -> [NSLayoutConstraint] {
        var constraints: [NSLayoutConstraint] = []
        let height = view != self.titleView ? self.minControlHeight : self.minTitleHeight
        if let sibling = sibling {
            let format = "V:[view(>=height)]-spacing-[sibling]"
            constraints.append(contentsOf: NSLayoutConstraint.constraints(withVisualFormat: format,
                options: NSLayoutFormatOptions(),
                metrics: ["spacing": self.spacing, "height": height],
                views: ["view": view, "sibling": sibling]) )
        } else {
            let format = "V:[view(>=height)]-0-|"
            constraints.append(contentsOf: NSLayoutConstraint.constraints(withVisualFormat: format,
                options: NSLayoutFormatOptions(),
                metrics: ["height": height],
                views: ["view": view]) )
        }
        if isLast {
            constraints.append(contentsOf: NSLayoutConstraint.constraints(withVisualFormat: "V:|-0-[view]",
                options: NSLayoutFormatOptions(),
                metrics: nil,
                views: ["view": view]) )
        }
        return constraints
    }

    private func updateUI() {
        self.view.backgroundColor = self.backgroundColor
        self.view.removeConstraints(self.view.constraints)
        self.setupGroupViews()
        self.updateContraints()
    }


    // MARK: Event handling

    @objc func handleTaps(_ sender: UIControl) {
        let index = sender.tag
        let action = self.actions[index]
        if action.enabled {
            self.disableControls()
            self.dismiss(animated: true, completion: { [unowned self] in
                self.enableControls()
                self.dismissCompletionBlock?()
                action.handler?(action)
            })
        }
    }

    func enableControls() {
        self.setUserInteractionOnControlsEnabled(true, controls: self.actionControls)
        self.setUserInteractionOnControlsEnabled(true, controls: self.cancelControls)
    }

    func disableControls() {
        self.setUserInteractionOnControlsEnabled(false, controls: self.actionControls)
        self.setUserInteractionOnControlsEnabled(false, controls: self.cancelControls)
    }

    func setUserInteractionOnControlsEnabled(_ enabled: Bool, controls: [UIControl]) {
        for control in controls {
            control.isUserInteractionEnabled = enabled
        }
    }

    @objc func recognizeGestures(_ gestureRecognizer: UIGestureRecognizer) {
        let point = gestureRecognizer.location(in: self.view)
        let view = self.view.hitTest(point, with: nil)
        if (view == self.view && self.outsideGestureShouldDismiss) {
            self.disableControls()
            self.dismiss(animated: true, completion: { [unowned self] in
                self.enableControls()
                self.dismissCompletionBlock?()
            })
        }
    }
}


// MARK: - Helpers for adding views

extension AFMActionSheetController {

    @objc
    open func add(_ action: AFMAction) {
        let control = UIButton.control(with: action)
        self.add(action, with: control)
    }

    @objc
    open func add(cancelling action: AFMAction) {
        let control = UIButton.control(with: action)
        self.add(cancelling: action, with: control)
    }

    @objc
    open func add(_ action: AFMAction, with control: UIControl) {
        self.add(action, with: control, andActionIsCancelling: false)
    }

    @objc
    open func add(cancelling action: AFMAction, with control: UIControl) {
        self.add(action, with: control, andActionIsCancelling: true)
    }

    @objc
    open func insert(_ action: AFMAction, at position: Int) {
        let control = UIButton.control(with: action)
        self.insert(action, with: control, at: position)
    }

    @objc
    open func insert(cancelling action: AFMAction, at position: Int) {
        let control = UIButton.control(with: action)
        self.insert(cancelling: action, with: control, at: position)
    }

    @objc
    open func insert(_ action: AFMAction, with control: UIControl, at position: Int) {
        self.insert(action, with: control, at: position, andActionIsCancelling: false)
    }

    @objc
    open func insert(cancelling action: AFMAction, with control: UIControl, at position: Int) {
        self.insert(action, with: control, at: position, andActionIsCancelling: true)
    }

    @objc
    open func add(titleLabelWith text: String) {
        let label = UILabel.title(with: text)
        self.add(title: label)
    }
}


// MARK: - Default control

extension UIButton {
    class func control(with action: AFMAction) -> UIButton {
        let button = UIButton()
        button.backgroundColor = UIColor.white
        button.setTitle(action.title, for: UIControlState())
        button.setTitleColor(UIColor.darkText, for: UIControlState())
        button.setTitleColor(UIColor.darkText.withAlphaComponent(0.5), for: .disabled)
        button.setTitleColor(UIColor.red, for: .highlighted)

        return button
    }
}

extension UILabel {
    class func title(with text: String) -> UILabel {
        let title = UILabel()
        title.text = text
        title.textAlignment = .center
        title.backgroundColor = UIColor.white

        return title
    }
}
